GetStream/stream-chat-android
# [4017] Make disconnection flow more errorless

### 🎯 Changes

- event handling in `ChatClient `is not behind a `Mutex`
- introduced new coroutine elements for `UserScope` such as `UserIdentifier` and `UserJob`

### 🧪 Testing

- apply the patch below 
- launch the XML sample app
- login by any user
- switch to another user
- repeat the steps above with different in-market `Socket` implementation

Make sure sample works with no errors or side-effects

<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt b/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt
--- a/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt	(revision f2b189332ae57e10eb11fadaeea7f8f3c95fc8a2)
+++ b/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt	(date 1662719795933)
@@ -191,6 +191,7 @@
 import io.getstream.logging.android.AndroidStreamLogger
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
 import kotlinx.coroutines.flow.MutableSharedFlow
 import kotlinx.coroutines.flow.first
 import kotlinx.coroutines.job
@@ -348,6 +349,7 @@
                 clientState.toMutableState()?.setConnectionState(ConnectionState.CONNECTING)
             }
             is DisconnectedEvent -> {
+                delay(1000L)
                 logger.i { "[handleEvent] event: DisconnectedEvent(disconnectCause=${event.disconnectCause})" }
                 when (event.disconnectCause) {
                     is DisconnectCause.ConnectionReleased,
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/user_login/UserLoginViewModel.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/user_login/UserLoginViewModel.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/user_login/UserLoginViewModel.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/user_login/UserLoginViewModel.kt	(revision f2b189332ae57e10eb11fadaeea7f8f3c95fc8a2)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/feature/user_login/UserLoginViewModel.kt	(date 1662719795914)
@@ -73,8 +73,8 @@
         ChatClient.instance().run {
             if (switchUser) {
                 switchUser(chatUser, user.token) {
-                    _events.postValue(Event(UiEvent.RedirectToChannels))
                 }.enqueue(::handleUserConnection)
+                _events.postValue(Event(UiEvent.RedirectToChannels))
             } else if (getCurrentUser() == null) {
                 connectUser(chatUser, user.token).enqueue(::handleUserConnection)
                 _events.postValue(Event(UiEvent.RedirectToChannels))
```

</details>


### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_
